### PR TITLE
#159568617 Design add new role modal

### PR DESCRIPTION
**Pull request by bahdcoder**
**What does this PR do?**

Designs a modal for adding a new role

**Description of Task to be completed?**

 - add ActionModal component for reusable modals
 - modify Checkbox component to emit change event
 - add UserRolesPage component
 - add styles for add new role modal and page

**How should this be manually tested?**

- Visit settings tab
- Click `New Role` button

**Any background context you want to provide?**

xxx

**What are the relevant pivotal tracker stories?**

[#159568617](https://www.pivotaltracker.com/story/show/159568617)
**Screenshots (if appropriate)**

screen shot 2018-08-09 at 7 09 24 am
screen shot 2018-08-09 at 7 09 47 am

**Questions:**

xxx

### CODE REVIEW

File: src/components/InputBox/index.tsx [hunk 1]
```diff
@@ -7,34 +7,44 @@ import './InputBox.scss';
...
-    />
+const InputBox = ({
+  error,
+  placeholder,
+  onChange,
+  value,
+  name,
+  textarea = false,
+}: InputBoxProps) => (
+  <div className={error ? 'input-box-group input-box-group--error' : 'input-box-group'}>
```
[the remark this answers is not in the document]
Comment (bahdcoder) [reply src/components/InputBox/index.tsx]: Thanks a lot for pointing this out. Was a rebase gone wrong. This has been resolved.

File: src/pages/CategoryPageForm/index.tsx [hunk 2]
```diff
@@ -390,6 +390,8 @@ export class CategoryPageForm extends React.Component<CategoryPageFormProps, Cat
                     enableConfigOptions={false}
                     isDisabled={true}
                     value={attribute.inputControl}
+                    // tslint:disable-next-line
+                    onAdditionalAttributeChange={() => {}}
```
Comment (andela-kakpobome) [inline src/pages/CategoryPageForm/index.tsx]: This does not need to be here, this use case makes it clear that this prop is optional

File: src/pages/UserRolesPage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        },
+      }) : role),
+    });
+  }
+
+  /**
+   * Creates the modal for adding roles
+   */
+  renderCreateRoleModal = () => ((
+    <ActionModal
```
Comment (andela-kakpobome) [inline src/pages/UserRolesPage/index.tsx]: What's the value of this abstraction as opposed to using the regular Modal component?

File: src/pages/UserRolesPage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                      </div>
+                    ))}
+                  </div>
+                ))}
+              </div>
+            </div>
+          </div>
+        </div>
+      )}
+      actionButtonName="ADD ROLE"
```
Comment (andela-kakpobome) [inline src/pages/UserRolesPage/index.tsx]: Nitpick, I think it'll be less `ignored` if this prop came before the renderContentProp

File: src/pages/UserRolesPage/index.tsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+   */
+  toggleModal = () => this.setState({
+    isModalOpen: !this.state.isModalOpen,
+  })
+
+  /**
+   * Handles input changes
+   */
+  // tslint:disable-next-line
+  handleFieldChange = () => {}
```
Comment (andela-kakpobome) [inline src/pages/UserRolesPage/index.tsx]: I don't want us to have unnecessary ignore lines, let this update state accordingly, so we don't need that ignore line/

File: src/components/ActionModal/ActionModal.test.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    actionButtonName: 'test-action-name',
+    showModal: true,
+  };
+  const wrapper = mount(<ActionModal { ...props } />);
+
+  it('passes down `classes` prop to the underlying modal', () => {
+    expect(wrapper.find(`.${props.classes}`).length).toBe(1);
+  });
+
+  it('class toggleModal function when cancel button is clicked', () => {
```
[the remark this answers is not in the document]
Comment (bahdcoder) [reply src/components/ActionModal/ActionModal.test.tsx]: Resolved.

File: src/components/ActionModal/ActionModal.test.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    wrapper.find({ name: 'CANCEL' }).first().simulate('click');
+    
+    expect(props.toggleModal).toHaveBeenCalled();
+  });
+
+  it('renders a button with name as the `actionButtonName` prop', () => {
+    expect(wrapper.find({ name: props.actionButtonName }).length).toBe(2);
```
[the remark this answers is not in the document]
Comment (bahdcoder) [reply src/components/ActionModal/ActionModal.test.tsx]: Test has been modified to fit description.

File: src/components/InputBox/index.tsx [hunk 8]
```diff
@@ -13,24 +13,37 @@ const InputBox = ({
...
+    {
+      !textarea ?
+        <input
+        type="text"
+        className={classes}
+        onChange={onChange}
+        value={value}
+        name={name}
+        placeholder={placeholder}
+        disabled={isDisabled}
```
[the remark this answers is not in the document]
Comment (bahdcoder) [reply src/components/InputBox/index.tsx]: Resolved.

File: src/pages/CategoryPageForm/index.tsx (hunk 2 above)
Comment (bahdcoder) [reply src/pages/CategoryPageForm/index.tsx]: Resolved.

File: src/pages/UserRolesPage/index.tsx (hunk 3 above)
Comment (bahdcoder) [reply src/pages/UserRolesPage/index.tsx]: The main aim of this is to refactor the css & mockup into one component. If you notice, we have about 6 - 7 modals of this exact same design (similar header and footer) in the application. If there's no common component, then each one will have a separate design, causing a repeat of the same mockup and css over and over. This component can be used as a foundation for refactoring.

File: src/pages/UserRolesPage/index.tsx (hunk 4 above)
Comment (bahdcoder) [reply src/pages/UserRolesPage/index.tsx]: Resolved.

File: src/pages/UserRolesPage/index.tsx (hunk 5 above)
Comment (bahdcoder) [reply src/pages/UserRolesPage/index.tsx]: Resolved.

File: src/pages/UserRolesPage/index.tsx (hunk 3 above)
Comment (andela-kakpobome) [reply src/pages/UserRolesPage/index.tsx]: So this modal is meant to replace DeleteModal, NotificationModal etc?

File: src/components/InputBox/InputBox.scss [hunk 9]
```diff
@@ -44,4 +44,16 @@
       color: map-get($palette, danger);
     }
   }
+
+  &__textarea-control {
+    @extend .input-box-group__control;
+    padding: rem(13px) rem(15px);
```
Comment (Billmike) [inline src/components/InputBox/InputBox.scss]: `width` and `height` before padding.

**Pull request merged**